Rate limit incoming packets from the client

Spottedleaf opened the pull request:
Packets are rate-limited through an interval of time to make sure latency
spikes don't kill clients. The default limit is 250 packets/second, which
is also far greater than the max ~70 packets/second vanilla will do according
to my testing.

Without these rate limits in place I can happily send 40,000 move packets per
second which will take down the server, as vanilla does zero rate limiting.

Clients violating the limit will be kicked and it will be logged to the console. The kick
message is configurable.

The rate limit can configured in paper.yml, and it can be disabled by being set to -1.

Spottedleaf committed:
Rate limit incoming packets from the client

Packets are ratelimited through an interval of time to make sure latency
spikes don't kill clients. The default limit is 250 packets/second, which
is also far greater than the max ~70 packets/second vanilla will do according
to my testing.

Without these ratelimits in place I can happily send 40,000 move packets per
second which will take down the server.

aikar commented on Spigot-Server-Patches/0424-Rate-limit-packets-incoming-from-players.patch:
```diff
@@ -0,0 +1,104 @@
+     }
+     // CraftBukkit end
++    // Paper start - ratelimit packets
++    private static final int PACKET_LIMIT_TIME_INTERVAL = 5; // seconds
```
we expect at least 20 per second from player updates... this feels tight to be a 5s window with a default of 250.

i think 1 second windows intervals is more intuitive and friendly with a default of 150 maybe

Spottedleaf replied:
The 5 seconds is just the sampling interval. 

The actual limit of packets is config * interval time, see
https://github.com/PaperMC/Paper/pull/1898/files#diff-94d6881867da7acaee9d07759b8ca37cR65

Spottedleaf replied:
The interval is large to accommodate large latency spikes for clients, although I'm going to change the default to 150 to better target peak packet rates.

aikar commented on Spigot-Server-Patches/0424-Rate-limit-packets-incoming-from-players.patch:
```diff
@@ -0,0 +1,104 @@
++        }
++
++        // Note: SeqLock would perform better here if needed
++        synchronized (this.rateLimitPacketsLock) {
```
we should be able to do this with Atomic Integers instead of synchronizing? 
other rate limit stuff is already done as atomic operations such as chat spam.

Spottedleaf replied:
That's because those methods have the main thread decrement them. If we rely on the main thread, then it's possible for the main thread to stall and for us to kick everyone. I don't think that's remotely desirable.

I could look into doing this lockless but the solution isn't pretty and less maintainable (at least if I want to avoid race conditions)

aikar commented on Spigot-Server-Patches/0424-Rate-limit-packets-incoming-from-players.patch:
```diff
@@ -0,0 +1,104 @@
++                        return; // no point, we've disconnected already
++                    }
++                    PlayerConnection.this.disconnect(com.destroystokyo.paper.PaperConfig.packetRateLimitKickMessage);
++                    PlayerConnection.LOGGER.warn("{} was kicked for sending too many packets!",
```
might be nice to print how many packets

Spottedleaf replied:
Added in next commit

aikar commented on Spigot-Server-Patches/0424-Rate-limit-packets-incoming-from-players.patch:
```diff
@@ -0,0 +1,104 @@
+     }
+ 
+     private static <T extends PacketListener> void a(Packet<T> packet, PacketListener packetlistener) {
++        // Paper start - Ratelimit packets
```
is this same method also called for sync processing? we only need to check in the async handler.

Spottedleaf replied:
This method is asynchronous. We want to prevent flooding of the main thread.

I'm also going to use this method in the next commit to stop packets altogether once we've made the decision to kill the connection.

Spottedleaf committed:
Changes on GH

- Lower default rate to 150packets/s from 250packets/
- Log how many packets the player was sending in the console (along with
the time interval for those packets)
- Packets are no longer processed once a client has been marked as rate
limiting
- Fix off-by-one error calculation

Spottedleaf committed:
Use a proper interval amongst other things

Increase the default limit back up to 250
Increase the default interval to 10 seconds

These numbers have been chosen since it allows for a client that
is sending 80 packets/s but lags for 15 seconds, and sends 80 packet/s
for another 10 seconds. The current code will read that as
80 * 15 + 80 * 10 = 2,000 packets in 10 seconds, which will be 200.0packets/s

With a default of 250, the client in this case will not be kicked.

However server owners might not care about this, so I've made the interval
configurable. They can lower it to limit peak packets/s.

Finally, a proper interval has been implemented. It will store packet counts
in buckets over a set interval. The sum of packets over the interval is cached
for performance as well.
